psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -888,24 +980,436 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+    std::vector<int> basis_endpoints_for_shell;
+
+    int atomic_ind = -1;
+    for (int P = 0; P < nshell; P++) {
+        if (primary_->shell(P).ncenter() > atomic_ind) {
+            shell_endpoints_for_atom.push_back(P);
+            atomic_ind++;
+        }
+        basis_endpoints_for_shell.push_back(primary_->shell_to_basis_function(P));
+    }
+    shell_endpoints_for_atom.push_back(nshell);
+    basis_endpoints_for_shell.push_back(nbf);
+
+    size_t natom = shell_endpoints_for_atom.size() - 1;
+
+    size_t max_functions_per_atom = 0L;
+    for (size_t atom = 0; atom < natom; atom++) {
+        size_t size = 0L;
+        for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+            size += primary_->shell(P).nfunction();
+        }
+        max_functions_per_atom = std::max(max_functions_per_atom, size);
+    }
+
+    if (debug_) {
+        outfile->Printf("  ==> LinK: Atom Blocking <==\n\n");
+        for (size_t atom = 0; atom < natom; atom++) {
+            outfile->Printf("  Atom: %3d, Atom Start: %4d, Atom End: %4d\n", atom, shell_endpoints_for_atom[atom],
+                            shell_endpoints_for_atom[atom + 1]);
+            for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+                int size = primary_->shell(P).nfunction();
+                int off = primary_->shell(P).function_index();
+                int off2 = basis_endpoints_for_shell[P];
+                outfile->Printf("    Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // ==> Prep Atom Pairs <== //
+    // Based on limited evidence, atom blocking is better for parallel performance than shell-pair blocking
+
+    std::vector<std::pair<int, int>> atom_pairs;
+    for (size_t Patom = 0; Patom < natom; Patom++) {
+        for (size_t Qatom = 0; Qatom <= Patom; Qatom++) {
+            bool found = false;
+            for (int P = shell_endpoints_for_atom[Patom]; P < shell_endpoints_for_atom[Patom + 1]; P++) {
+                for (int Q = shell_endpoints_for_atom[Qatom]; Q < shell_endpoints_for_atom[Qatom + 1]; Q++) {
+                    if (ints[0]->shell_pair_significant(P, Q)) {
+                        found = true;
+                        atom_pairs.emplace_back(Patom, Qatom);
+                        break;
+                    }
+                }
+                if (found) break;
+            }
+        }
+    }
+
+    // ==> Prep Bra-Bra Shell Pairs <== //
+
+    // A comparator used for sorting integral screening values
+    auto screen_compare = [](const std::pair<int, double> &a, 
+                                    const std::pair<int, double> &b) { return a.second > b.second; };
+
+    std::vector<std::vector<int>> significant_bras(nshell);
+    double max_integral = ints[0]->max_integral();
+
+#pragma omp parallel for
```

Comment:
I notice that, with the parallelization of the LinK preprocessing steps being introduced, there are four separate instances of "#pragma omp parallel for" being used in the LinK K build. As this constitutes four separate OpenMP threaded regions, this would incur an overhead of four OpenMP thread forks and four OpenMP thread joins per SCF iteration, just from the call to LinK. 

With a bit of refactoring, it's likely possible to encompass all current parallel LinK preprocessing loops, as well as the core LinK parallel loop itself, into a single "#pragma omp parallel" region, with each loop having its own "#pragma omp for" associated with it. This would achieve the same parallelization effect as currently, but using only a single OpenMP parallel region, thus requiring only a single OpenMP thread fork/join per LinK call.

This is something of a micro-optimization, though, so I wonder if it is worth the time spent refactoring.

Reply:
This sounds like something better left to a follow-up PR.

Reply:
That's reasonable. I figured I would at least bring up the suggestion, though.